src/asm2wasm.h

```diff
@@ -206,6 +218,85 @@ struct Asm2WasmPreProcessor {
       *marker = START_FUNCS[0];
     }
 
+    // handle debug info, if this build wants that.
+    if (debugInfo) {
+      // asm.js debug info comments look like
+      //   ..command..; //@line 4 "tests/hello_world.c"
+      // we convert those into emscripten_debuginfo(file, line)
+      // calls, where the params are indices into a mapping. then
+      // the compiler and optimizer can operate on them. after
+      // that, we can apply the debug info to the wasm node right
+      // before it - this is guaranteed to be correct without opts,
+      // and is usually decently accurate with them.
+      const auto SCALE_FACTOR = 1.25; // an upper bound on how much more space we need as a multiple of the original
+      const auto ADD_FACTOR = 100; // an upper bound on how much we write for each debug info element itself
+      auto size = strlen(input);
+      auto upperBound = Index(size * SCALE_FACTOR) + ADD_FACTOR;
+      char* copy = allocatedCopy = (char*)malloc(upperBound);
+      char* end = copy + upperBound;
+      char* out = copy;
+      std::string DEBUGINFO_INTRINSIC = EMSCRIPTEN_DEBUGINFO.str;
+      auto DEBUGINFO_INTRINSIC_SIZE = DEBUGINFO_INTRINSIC.size();
+      bool seenUseAsm = false;
+      auto startsWith = [&](const char *s) {
```

Comment:
I would make this a top-level function as opposed to a lambda

1) when reading this code it can be assumed that `startsWith` exists elsewhere, so it doesn't interrupt the flow of reading this function
2) the way its currently implemented a better name would be `inputAtTheCurrentPositionStartsWith("foo")`. I feel `startsWith(input, "foo")` is more explicit at the call site and requires less context for the reader to keep in their head

Both should be inlineable by the compiler so it should be equivalent to the unrolled version, either way.